agent/engine/host_resource_manager.go

```diff
@@ -16,20 +16,226 @@
+func (h *HostResourceManager) consume(taskArn string, resources map[string]*ecs.Resource) (bool, error) {
+	h.hostResourceManagerRWLock.Lock()
+	defer h.hostResourceManagerRWLock.Unlock()
+
+	defer logger.Debug("Consumed resources after task consume call", logger.Fields{
+		"taskArn":   taskArn,
+		"CPU":       *h.consumedResource["CPU"].IntegerValue,
+		"MEMORY":    *h.consumedResource["MEMORY"].IntegerValue,
+		"PORTS_TCP": h.consumedResource["PORTS_TCP"].StringSetValue,
+		"PORTS_UDP": h.consumedResource["PORTS_UDP"].StringSetValue,
+		"GPU":       *h.consumedResource["GPU"].IntegerValue,
+	})
+
+	// Check if already consumed
+	_, ok := h.consumedResource[taskArn]
+	if ok {
+		// Nothing to do, already consumed, return
+		return true, nil
+	}
+
+	ok, err := h.consumable(resources)
+	if err != nil {
+		return false, err
+	}
+	if ok {
+		// CPU
+		cpu := *h.consumedResource["CPU"].IntegerValue + *resources["CPU"].IntegerValue
+		h.consumedResource["CPU"].SetIntegerValue(cpu)
+
+		// MEM
+		mem := *h.consumedResource["MEMORY"].IntegerValue + *resources["MEMORY"].IntegerValue
+		h.consumedResource["MEMORY"].SetIntegerValue(mem)
+
+		// PORTS
+		portsResource, ok := resources["PORTS_TCP"]
+		if ok {
+			taskPortsSlice := portsResource.StringSetValue
+			for _, port := range taskPortsSlice {
+				// Create a copy to assign it back as "PORTS_TCP"
+				newPortResource := h.consumedResource["PORTS_TCP"]
+				newPorts := append(h.consumedResource["PORTS_TCP"].StringSetValue, port)
+				newPortResource.StringSetValue = newPorts
+				h.consumedResource["PORTS_TCP"] = newPortResource
+			}
+		}
+
+		// PORTS_UDP
+		portsResource, ok = resources["PORTS_UDP"]
+		if ok {
+			taskPortsSlice := portsResource.StringSetValue
+			for _, port := range taskPortsSlice {
+				newPortResource := h.consumedResource["PORTS_UDP"]
+				newPorts := append(h.consumedResource["PORTS_UDP"].StringSetValue, port)
+				newPortResource.StringSetValue = newPorts
+				h.consumedResource["PORTS_UDP"] = newPortResource
+			}
+		}
+
+		// GPU
+		*h.consumedResource["GPU"].IntegerValue += *resources["GPU"].IntegerValue
+
+		// Set consumed status
+		h.taskConsumed[taskArn] = true
+		return true, nil
+	}
+	return false, nil
+}
+
+// Helper function for consume to check if resources are consumable with the current account
+// we have for the host resources. Should not call host resource manager lock in this func
+// return values
+func (h *HostResourceManager) consumable(resources map[string]*ecs.Resource) (bool, error) {
+	// CPU
+	cpuResource, ok := resources["CPU"]
+	if ok {
+		if *(h.initialHostResource["CPU"].IntegerValue) < *(h.consumedResource["CPU"].IntegerValue)+*(cpuResource.IntegerValue) {
+			return false, nil
+		}
+	} else {
+		return false, fmt.Errorf("No CPU in task resources")
+	}
+
+	// MEM
+	memResource, ok := resources["MEMORY"]
+	if ok {
+		if *(h.initialHostResource["MEMORY"].IntegerValue) < *(h.consumedResource["MEMORY"].IntegerValue)+*(memResource.IntegerValue) {
```

Comment:
Although it will be guaranteed as we will only feed output of task.ToHostResources(#3706), I think it can be good to have a resourceMap health check function implemented. Let me do that